common/datavalues/src/columns/primitive/mod.rs

```diff
@@ -190,34 +191,49 @@ impl<T: PrimitiveType> Column for PrimitiveColumn<T> {
             return Arc::new(self.clone());
         }
 
-        let mut res = Vec::<T>::with_capacity(length);
-        let mut offset = 0;
-        let values = self.values();
-
-        const MASK_BITS: usize = 64;
-        for mut mask in filter.values().chunks::<u64>() {
-            if mask == u64::MAX {
-                res.extend(&values[offset..offset + MASK_BITS]);
-            } else {
-                while mask != 0 {
-                    let n = std::intrinsics::cttz(mask) as usize;
-                    res.push(values[offset + n]);
-                    mask = mask & (mask - 1);
+        let mut new = Vec::<T>::with_capacity(length);
+        let mut dst = new.as_mut_ptr();
+
+        const CHUNK_SIZE: usize = 64;
+        let mut chunks = self.values().chunks_exact(CHUNK_SIZE);
+        let mut mask_chunks = filter.values().chunks::<u64>();
```

Comment:
There is a faster path for `mask_chunks`, if the filter has no offsets, it could use `chunks_exact` to have better performance.  See: https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/filter.rs#L142-L147

There are two approaches:

1.  Use generic dispatch like `BitChunkIterExact`

2. if offset > 0, use an extra loop to consume the offset, this is called `Header Loops`, then we can continue the  `Main Loops` && `Tail Loops`.

The second one may have better performance because `chunks.iter()` must merge two u64 values to generate one merged u64, see:  merge_reversed in `arrow2`